server-spi-private/src/main/java/org/keycloak/validate/Validator.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.validate;
+
+import org.keycloak.provider.Provider;
+
+import java.util.Map;
+
+/**
+ * Validates given input in a {@link ValidationContext}.
+ * <p>
+ * Validations can be supported with an optional {@code inputHint}, which could denote a reference to a potentially
+ * nested attribute of an object to validate.
+ * <p>
+ * Validations can be configured with an optional {@code config} {@link Map}.
+ */
+public interface Validator extends Provider {
+
+    /**
+     * Validates the given {@code input}.
+     *
+     * @param input the value to validate
+     * @return the validation context with the outcome of the validation
+     */
+    default ValidationContext validate(Object input) {
+        return validate(input, "input", new ValidationContext(), ValidatorConfig.EMPTY);
+    }
+
+    /**
+     * Validates the given {@code input} with an additional {@code config}.
+     *
+     * @param input  the value to validate
+     * @param config parameterization for the current validation
+     * @return the validation context with the outcome of the validation
+     */
+    default ValidationContext validate(Object input, ValidatorConfig config) {
```

Comment:
1. Was it considered to have ValidatorConfig as part of the ValidationContext?
2. The ValidationContext including session should always be available to the validator, e.g for getting other validators and delegate. 
But method `ValidationContext validate(Object input)` allows to request validation without session available. Waht is a validotor supposed to do which needs the session?

Reply:
1. Yes it was, I had this (via simple `Map<String,Object> attributes`) in my initial API design: https://github.com/keycloak/keycloak/pull/7324/files#diff-a30ff7afb40dfeee146a59e5e46fbd15598d5d36a2dbd616095d049e5181613dR36
We want to support simple and complex validations. Simple validations (just one call to validate(...)) don't require an initial `ValidationConfig` but might want to fine tune the `Validator`.
Complex validations might invoke `validate(...)` consecutively for different `Validator`s with a shared `ValidationContext`. Since the `ValidationContext` is shared across multiple `Validator` invocations, it would require additional context management to maintain the proper `ValidationConfig`. Explicitly passing the `ValidationConfig` makes the API easier for simple and complex validations.

2. A `Validator` which needs the `KeycloakSession` needs to be invoked with a properly configured `ValidationContext` with a session. `#validate(object)` allows simple validations like
```
Assert.assertTrue(Validators.email().validate("admin@example.org").isValid());
```
while still sharing the common `Validator` interface.

Reply:
1. Thanks, shared context but not shared config is a good reason to separate the parameters
2. Ok, got the different cases. The main drawback is the difficulty for the validator implementation to state that that  a 'ValidationContext' instance is required and must not be null.  Would be nice to know that just by the interface. 
```
public interface Validator {
    ValidationContext validate(Object input, ValidatorConfig config)) ...

public interface ValidatorWithContext {
    ValidationContext validate(Object input, ValidationContext context, ValidatorConfig config)) ...
```
Unfortunately this can't be done for 'ValidatorConfig' parameter at the same time. Get's a bit unhandy to have a 'ValidatorWithContextAndConfig' and a 'ValidatorWithConfig' interface on top.